pymodelextractor/learners/observation_table_learners/generic_lstar_learner.py

```diff
@@ -0,0 +1,239 @@
+        start_time = time.time()
+        teacher.log_hierachy = log_hierachy
+        self.log_hierachy = log_hierachy
+        if self.log_hierachy != no_log:
+            print("**** Started lstar learning ****")
+        self._teacher = teacher
+        self._symbols = self._teacher.alphabet.symbols
+        self._build_observation_table()
+        self._initialize_observation_table()
+        model = None
+        answer = False
+        counter = 1
+        counterexample_counter = 0
+
+        while not answer:
+            start_iteration_time = time.time()
+            if self.log_hierachy == debug_log or self.log_hierachy == trace_log:
+                print(" # Starting iteration " + str(counter))
+            surpassed_max_query_len = self._close()
+            
+            if surpassed_max_query_len:
+                return self._learning_results_for(self._history[-1] if len(self._history) > 0 else None, time.time() - start_time)
+
+            surpassed_max_query_len = self._make_consistent()
+            
+            if surpassed_max_query_len:
+                return self._learning_results_for(self._history[-1] if len(self._history) > 0 else None, time.time() - start_time)
+
+            self._model_translator._output_alphabet = self._teacher.output_alphabet
+            model = self._model_translator.translate(
+                self._observation_table, self._teacher.alphabet, self._teacher.output_alphabet)
+            self._history.append(model)
+            
+            if self._max_time is not None: 
+                return_dict[self._max_time] = self._learning_results_for(self._history[-1] if len(self._history) > 0 else None, None)
+
+            start_eq_time = time.time()
+            answer, counterexample = self._teacher.equivalence_query(model)
+            eq_duration = time.time() - start_eq_time
+
+            if (self._max_states is not None) and (len(model.states) > self._max_states):
+                return self._learning_results_for(model, time.time() - start_time)
+
+            if not answer:
+                if self.log_hierachy == debug_log or self.log_hierachy == trace_log:
+                    print("    - Found counterexample in " + str(eq_duration) + "s -> " + str(counterexample))
+                counterexample_counter += 1
+
+                surpassed_max_query_len = self._update_observation_table_with(counterexample)
+                if surpassed_max_query_len:
+                    return self._learning_results_for(self._history[-1] if len(self._history) > 0 else None, time.time() - start_time)
+                
+            else:
+                if self.log_hierachy == debug_log or self.log_hierachy == trace_log:
+                    print("    - Made equivalence query in " + str(eq_duration) + "s")
+
+            duration = time.time() - start_iteration_time
+            if self.log_hierachy == debug_log or self.log_hierachy == trace_log:
```

Comment:
debug levels should be numeric, and you should compare with `>` isntead of every single level

Reply:
or look into pythons native log package (has it's own levels)